sdk/clientcore/core/src/main/java/io/clientcore/core/http/pipeline/HttpPipelineBuilder.java

```diff
@@ -90,20 +131,411 @@ public HttpPipelineBuilder httpClient(HttpClient httpClient) {
+     * @throws IllegalStateException If the name of the {@code policy} is reserved.
+     */
+    public HttpPipelineBuilder setPolicy(HttpPipelinePolicy policy) {
+        Objects.requireNonNull(policy, "'policy' cannot be null.");
+        String lowerCaseName = policy.getName().toLowerCase();
+        if (!policyNames.contains(lowerCaseName)) {
+            throw LOGGER.logThrowableAsError(
+                new NoSuchElementException("A policy with the name '" + policy.getName() + "' was not found."));
+        }
+
+        if (RESERVED.contains(lowerCaseName)) {
+            throw LOGGER
+                .logThrowableAsError(new IllegalStateException("The policy name is reserved: " + policy.getName()));
+        }
+
+        return setOrRemove(policy, policy.getName());
+    }
+
+    /**
+     * Removes the policy with the given {@code policyName} from the pipeline.
+     * <p>
+     * If a policy with the given name is not found this is a no-op.
+     * <p>
+     * If the {@code policyName} is any of {@code "redirect"}, {@code "retry"}, {@code "credential"}, {@code "logging"},
+     * or {@code "telemetry"} an {@link IllegalStateException} will be thrown. These names are reserved for the set
+     * methods that configure the respective policies.
+     *
+     * @param policyName The name of the policy to remove.
+     * @return The updated HttpPipelineBuilder object.
+     * @throws NullPointerException If {@code policyName} is null.
+     * @throws IllegalStateException If the {@code policyName} is reserved.
+     */
+    public HttpPipelineBuilder removePolicy(String policyName) {
+        Objects.requireNonNull(policyName, "'policyName' cannot be null.");
+        String lowerCaseName = policyName.toLowerCase();
+        if (!policyNames.contains(lowerCaseName)) {
+            return this;
+        }
+
+        if (RESERVED.contains(lowerCaseName)) {
+            throw LOGGER.logThrowableAsError(new IllegalStateException("The policy name is reserved: " + policyName));
+        }
+
+        return setOrRemove(null, policyName);
+    }
+
+    private HttpPipelineBuilder setOrRemove(HttpPipelinePolicy policy, String policyName) {
+        if (attemptToSetOrRemove(policy, policyName, beforeRedirect, afterCredential)) {
+            return this;
+        }
+
+        if (attemptToSetOrRemove(policy, policyName, beforeRetry, afterRetry)) {
+            return this;
+        }
+
+        if (attemptToSetOrRemove(policy, policyName, beforeCredential, afterCredential)) {
+            return this;
+        }
+
+        if (attemptToSetOrRemove(policy, policyName, beforeInstrumentation, afterInstrumentation)) {
+            return this;
+        }
+
+        throw LOGGER.logThrowableAsError(new IllegalStateException("Should not reach here."));
+    }
+
+    private boolean attemptToSetOrRemove(HttpPipelinePolicy policy, String policyName,
+        LinkedList<HttpPipelinePolicy> beforePolicies, LinkedList<HttpPipelinePolicy> afterPolicies) {
```

Comment:
We can get the `policyName` from the `policy` itself.

```suggestion
    private boolean attemptToSetOrRemove(HttpPipelinePolicy policy,
        LinkedList<HttpPipelinePolicy> beforePolicies, LinkedList<HttpPipelinePolicy> afterPolicies) {
```